File: Java-Kotlin/src/main/java/jianzhi/jz25.java
package jianzhi;

import common.datastructure.RandomListNode;

public class jz25 {

    public RandomListNode Clone(RandomListNode pHead) {
        if (pHead == null) {
            return null;
        }

        RandomListNode currentNode = pHead;
        while (currentNode != null) {
            RandomListNode copyNode = new RandomListNode(currentNode.label);

            RandomListNode next = currentNode.next;
            currentNode.next = copyNode;
            copyNode.next = next;
            currentNode = next;
        }

        currentNode = pHead;
        while (currentNode != null) {
            RandomListNode copyNode = currentNode.next;
            copyNode.random = currentNode.random == null ? null : currentNode.random.next;
            currentNode = copyNode.next;
        }

        RandomListNode clonedHead = pHead.next;
        currentNode = pHead;
        while (currentNode != null) {
            RandomListNode clonedNode = currentNode.next;
            currentNode.next = clonedNode.next;
            clonedNode.next = clonedNode.next == null ? null : clonedNode.next.next;

            currentNode = currentNode.next;
        }

        return clonedHead;
    }
}

File: Go-Solutions/nowcoder/jz25.go
package nowcoder

func Clone(head *RandomListNode) *RandomListNode {
	//write your code here
	if head == nil {
		return head
	}

	// insert a node after each node in the head
	headRunner := head
	//1、复制每个结点，如复制结点A得到A1，将结点A1插到结点A后面；
	for headRunner != nil {
		next := headRunner.Next

		copyNode := &RandomListNode{
			Label:  headRunner.Label,
			Next:   nil,
			Random: nil,
		}

		copyNode.Next = next
		headRunner.Next = copyNode
		headRunner = next
	}

	headRunner = head
	for headRunner != nil {
		copyNode := headRunner.Next
		if headRunner.Random == nil {
			copyNode.Random = nil
		} else {
			copyNode.Random = headRunner.Random.Next
		}
		headRunner = headRunner.Next.Next
	}

	headRunner = head
	cloneHead := head.Next
	for headRunner != nil {
		clonedNode := headRunner.Next
		headRunner.Next = clonedNode.Next
		if clonedNode.Next == nil {
			clonedNode.Next = nil
		} else {
			clonedNode.Next = clonedNode.Next.Next
		}
		headRunner = headRunner.Next
	}

	return cloneHead
}
